src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs

```diff
@@ -355,6 +377,47 @@ public void HandleLogMessage(TestMessageLevel level, string message)
         /// <param name="rawMessage">Raw message.</param>
         public void HandleRawMessage(string rawMessage)
         {
+            if (this.requestData.IsTelemetryOptedIn)
+            {
+                var message = this.dataSerializer.DeserializeMessage(rawMessage);
+
+                if (string.Equals(message.MessageType, MessageType.DiscoveryComplete))
+                {
+                    var discoveryCompletePayload =
+                        this.dataSerializer.DeserializePayload<DiscoveryCompletePayload>(message);
+
+                    if (discoveryCompletePayload != null)
+                    {
+                        if (discoveryCompletePayload.Metrics == null)
+                        {
```

Comment:
is this a sequential call i.e. can there be race?

Reply:
RawMessage of type DiscoveryCompletePayload is sent only once per Discovery Request. 
We are creating new instance of DiscoveryCompletePayload by deserializing the raw message.
So, I don't think this will result in an race condition.